Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,188 @@
+#import "WebExtensionScriptInjectionResultParameters.h"
+#import "WebExtensionDynamicScripts.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionTabIdentifier.h"
+#import "WebExtensionUtilities.h"
+#import "WebPageProxy.h"
+#import "WebProcessProxy.h"
+#import "WebUserContentControllerProxy.h"
+#import "_WKFrameHandle.h"
+#import "_WKFrameTreeNode.h"
+#import <WebCore/UserStyleSheet.h>
+#import <wtf/BlockPtr.h>
+#import <wtf/cocoa/VectorCocoa.h>
+
+namespace WebKit {
+using namespace DynamicScripts;
+
+using SourcePair = DynamicScripts::SourcePair;
+using ScriptVector = DynamicScripts::StyleSheetPairs;
+
+void WebExtensionContext::scriptingExecuteScript(const WebExtensionScriptInjectionParameters& parameters, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
+{
+    WKWebView *webView = m_tabMap.get(parameters.identifier.value())->mainWebView();
+    if (!hasPermission(webView.URL)) {
+        callback(std::nullopt);
+        return;
+    }
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, parameters = WTFMove(parameters), callback = WTFMove(callback)](_WKFrameTreeNode *mainFrame) mutable {
+        API::ContentWorld& executionWorld = *m_contentScriptWorld;
+        if (parameters.world == WebExtensionContentWorldType::Main)
+            executionWorld = API::ContentWorld::pageContentWorld();
+
+        Vector<RetainPtr<_WKFrameTreeNode>> frames;
+        if (parameters.allFrames || !parameters.frameIDs)
+            populateFrames(mainFrame, mainFrameID, parameters.allFrames, false, WTFMove(frames));
+        else
+            populateFrames(mainFrame, parameters.frameIDs.value(), WTFMove(frames));
+
+        size_t numberOfExecutions = frames.size();
+        Vector<WebExtensionScriptInjectionResultParameters> results;
+
+        for (auto& frame : frames) {
+            WKFrameInfo *info = frame.get().info;
+            NSURL *frameURL = info.request.URL;
+
+            if (!hasPermission(frameURL)) {
+                addResultAndCompleteIfDone(nil, info, toErrorString(@"scripting.executeScript()", nil, @"Extension does not have access to URL: %@", frameURL.absoluteString), numberOfExecutions, WTFMove(results), WTFMove(callback));
+                continue;
+            }
+
+            if (parameters.files) {
+                Vector<std::optional<SourcePair>> scriptPairs;
+                for (auto& file : parameters.files.value())
+                    scriptPairs.append(sourcePairForResource(file, m_extension));
+                
+                for (auto& script : scriptPairs) {
+                    [webView _evaluateJavaScript:script.value().first withSourceURL:script.value().second.value() inFrame:info inContentWorld:executionWorld.wrapper() completionHandler:makeBlockPtr([info = RetainPtr { info }, numberOfExecutions, results = WTFMove(results), callback = WTFMove(callback)](id resultOfJSExecution, NSError *error) mutable {
```

Comment:
This is also suspicious. You call `_evaluateJavaScript` for each element in `scriptPairs`, but you are using `WTFMove` on the callback each loop. That doesn't seem right.